src/main/java/org/traccar/protocol/LaipacProtocolDecoder.java

```diff
@@ -145,23 +239,39 @@ protected Object decode(
             position.setNetwork(new Network(CellTower.from(mcc, mnc, lac, cid)));
         }
 
+        String unused = parser.next();
+
         String checksum = parser.next();
 
         if (channel != null) {
-            if (event.equals("3")) {
-                channel.writeAndFlush(new NetworkMessage("$AVCFG,00000000,d*31\r\n", remoteAddress));
-            } else if (event.equals("X") || event.equals("4")) {
-                channel.writeAndFlush(new NetworkMessage("$AVCFG,00000000,x*2D\r\n", remoteAddress));
-            } else if (event.equals("Z")) {
-                channel.writeAndFlush(new NetworkMessage("$AVCFG,00000000,z*2F\r\n", remoteAddress));
-            } else if (Character.isLowerCase(status.charAt(0))) {
-                String response = "$EAVACK," + event + "," + checksum;
-                response += Checksum.nmea(response) + "\r\n";
-                channel.writeAndFlush(new NetworkMessage(response, remoteAddress));
-            }
+
+            sendAcknowledge(status, event, checksum, channel, remoteAddress);
+
+            String devicePassword = getDevicePassword(deviceSession);
+            sendEventResponse(event, devicePassword, channel, remoteAddress);
         }
 
         return position;
     }
 
+    @Override
+    protected Object decode(
+        Channel channel, SocketAddress remoteAddress, Object msg) throws Exception {
+
+        String sentence = (String) msg;
+
+        Parser parser = new Parser(PATTERN_ECHK, sentence);
```

Comment:
Looks like you can just check header using `startsWith` instead of trying to match all patterns.

Reply:
Each regex processing will stop when the beginning of the pattern doesn't match (almost the same, I think, that if we use `startsWith`). The only regex that will be fully parsed, and will be the last one, is the one that fully matches. In this case both code will be executed, `startsWith` code and regex matching code. I can be wrong, of course, I'm not an experienced Java programmer. If you don't agree, no problem, I'll do it your way. Thanks.

Reply:
Regex matching is still much slower than `startsWith` even if it fails early. Also, we do the same logic in other places, so it would be nice to have consistent implementations.

Reply:
I see your point. I'll do it that way then. Thanks.